src/Compilers/CSharp/Portable/Symbols/Source/ParameterHelpers.cs

```diff
@@ -21,16 +22,113 @@ internal static class ParameterHelpers
+                diagnostics,
+                allowRefOrOut,
+                allowThis,
+                addRefReadOnlyModifier,
+                suppressUseSiteDiagnostics: false,
+                lastIndex: syntax.Parameters.Count - 1,
+                parameterCreationFunc: (Binder context, Symbol owner, TypeWithAnnotations parameterType,
+                                        ParameterSyntax syntax, RefKind refKind, int ordinal,
+                                        SyntaxToken paramsKeyword, SyntaxToken thisKeyword, bool addRefReadOnlyModifier,
+                                        DiagnosticBag declarationDiagnostics) =>
+                {
+                    return SourceParameterSymbol.Create(
+                        context,
+                        owner,
+                        parameterType,
+                        syntax,
+                        refKind,
+                        syntax.Identifier,
+                        ordinal,
+                        (paramsKeyword.Kind() != SyntaxKind.None),
+                        ordinal == 0 && thisKeyword.Kind() != SyntaxKind.None,
+                        addRefReadOnlyModifier,
+                        declarationDiagnostics);
+                }
+);
+        }
+
+        public static ImmutableArray<FunctionPointerParameterSymbol> MakeFunctionPointerParameters(
+            Binder binder,
+            FunctionPointerMethodSymbol owner,
+            SeparatedSyntaxList<ParameterSyntax> parametersList,
+            DiagnosticBag diagnostics,
+            bool suppressUseSiteDiagnostics)
+        {
+            return MakeParameters(
+                binder,
+                owner,
+                parametersList,
+                out _,
+                diagnostics,
+                allowRefOrOut: true,
+                allowThis: false,
+                addRefReadOnlyModifier: true,
+                suppressUseSiteDiagnostics,
+                parametersList.Count - 2,
+                parameterCreationFunc: (Binder binder, FunctionPointerMethodSymbol owner, TypeWithAnnotations parameterType,
+                                        ParameterSyntax syntax, RefKind refKind, int ordinal,
+                                        SyntaxToken paramsKeyword, SyntaxToken thisKeyword, bool addRefReadOnlyModifier,
+                                        DiagnosticBag diagnostics) =>
+                {
+
+                    var locations = ImmutableArray.Create<Location>(new SourceLocation(syntax.Type));
+                    var modifierType = binder.GetWellKnownType(WellKnownType.System_Runtime_InteropServices_InAttribute, diagnostics, syntax);
+                    var customModifiers = addRefReadOnlyModifier && refKind == RefKind.In ?
+                        ImmutableArray.Create(CSharpCustomModifier.CreateRequired(modifierType)) :
+                        ImmutableArray<CustomModifier>.Empty;
+
+                    if (parameterType.IsVoidType())
```

Comment:
>if (parameterType.IsVoidType()) [](start = 20, length = 31)

Can we also share this logic with other parameters built from source?

Reply:
Other parameters report this at parsing time, because they don't have to know whether the current element is a return type that allows void, or a parameter type. Function pointers don't know whether it's a parameter until after parsing is done, so we have to do this at a later step.

---
In reply to: [362939946](https://github.com/dotnet/roslyn/pull/40431#discussion_r362939946) [](ancestors = 362939946)